File: leetcode_js/tests/helpers/treeNode.js
class TreeNode {
  constructor(val, left, right) {
    this.val = typeof val === 'undefined' ? 0 : val;
    this.left = typeof left === 'undefined' ? null : left;
    this.right = typeof right === 'undefined' ? null : right;
  }
}

const serialize = (root) => {
  if (root === null) {
    return [];
  }

  const result = [];
  const queue = [root];

  while (queue.length > 0) {
    const qlen = queue.length;
    for (let i = 0; i < qlen; i++) {
      const node = queue.shift();

      if (node === null) {
        result.push(null);
      } else {
        result.push(node.val);

        queue.push(node.left);
        queue.push(node.right);
      }
    }
  }

  while (result[result.length - 1] === null) {
    result.pop();
  }

  return result;
};

const deserialize = (arr) => {
  if (arr.length === 0) {
    return null;
  }

  const root = new TreeNode(arr[0]);
  const queue = [root];

  let i = 1;
  while (queue.length > 0 && i < arr.length) {
    const node = queue.shift();

    if (arr[i] !== null) {
      const left = new TreeNode(arr[i]);
      node.left = left;
      queue.push(left);
    }

    i++;

    if (arr[i] !== null) {
      const right = new TreeNode(arr[i]);
      node.right = right;
      queue.push(right);
    }

    i++;
  }

  return root;
};

module.exports = { TreeNode, serialize, deserialize };

File: leetcode_python/lib/helpers/tree_node.py
from typing import List, TypeVar, Generic, Union


T = TypeVar("T")


class TreeNode(Generic[T]):
    def __init__(self, val: T, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def to_list(self) -> List[T]:
        result = []
        queue = [self]

        while len(queue) > 0:
            size = len(queue)
            for i in range(size):
                node = queue[i]

                if node == None:
                    result.append(None)
                else:
                    result.append(node.val)
                    queue.append(queue, node.left, node.right)

            queue = queue[size:]

        i = len(result)
        while i > 0 and result[i - 1] == None:
            i -= 1

        return result[:i]

    @staticmethod
    def from_list(vals: List[Union[T, None]]) -> "TreeNode[T]":
        if len(vals) == 0:
            return None

        root = TreeNode(vals[0])
        queue = [root]

        i = 1
        while i < len(vals):
            node = queue[0]
            queue.pop(0)

            if i < n and vals[i] != None:
                node.left = TreeNode(vals[i])
                queue.append(node.left)

            i += 1

            if i < n and vals[i] != None:
                node.right = TreeNode(vals[i])
                queue.append(node.right)

            i += 1

        return root

    # def __eq__(self, other: "TreeNode[T]"):
    #     while self and other and self.val == other.val:
    #         self = self.left
    #         other = other.left

    #     if not self and not other:
    #         return True

    #     return False
